### ISSUE TITLE
NPE in DTLSConnector

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#1273]
We face some NPE with Cf-2.0.0-RC1.
```
Unexpected error occurred while processing record from peer [/52.187.235.18:59337]
java.lang.NullPointerException: null
 at org.eclipse.californium.scandium.dtls.Handshaker.processMessage(Handshaker.java:620) ~[lwm2mfe.jar:na]
 at org.eclipse.californium.scandium.DTLSConnector.processHandshakeRecord(DTLSConnector.java:1576) ~[lwm2mfe.jar:na]
 at org.eclipse.californium.scandium.DTLSConnector.processRecord(DTLSConnector.java:1289) ~[lwm2mfe.jar:na]
 at org.eclipse.californium.scandium.DTLSConnector$11.run(DTLSConnector.java:1158) [lwm2mfe.jar:na]
 at org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:276) [lwm2mfe.jar:na]
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212]
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212]
 at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]
```

It seems that the flight variable is NULL : https://github.com/eclipse/californium/blob/2.2.0-RC1/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java#L620

**Comment by boaks** [comment]
Do you know, if the handshake was completed or has failed?

**Comment by sbernard31** [comment]
nope :/

**Comment by boaks** [comment]
OK.

currently I see two suspects:

- some code violates the serial execution and terminates handshakes.
  (e.g. [processNewClientHello](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java#L1691)

- the handshake message, which triggers the NPE is from the `inboundMessageBuffer`
  [inboundMessageBuffer](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java#L655)

But for both, I don't see/understand, that this actually happens. I would start to check the ownership, when the handshake is failed or completed. We will see ...

**Comment by sbernard31** [comment]
Thx :pray:

I haven't much more information to provide for now.
This happens  ~30 time a days. (so not so much)
And we see that since we integrated Leshan-1.0.0-RC1(Cf-2.0.0-RC1). With Leshan-1.0.0-M13(Cf-2.0.0) it seems this did not happen.

**Comment by boaks** [comment]
I added some logging in #1274 

I think, the next step would be to create some "retransmissions in handshakes".

**Comment by boaks** [comment]
> And we see that since we integrated Leshan-1.0.0-RC1(Cf-2.0.0-RC1). With Leshan-1.0.0-M13(Cf-2.0.0) it seems this did not happen.

I'm now wondering. Is it important to know, if between 2.0.0-RC1 and 2.0.0 the issue maybe fixed?

**Comment by sbernard31** [comment]
Oups sorry typo : 

And we see that since we integrated Leshan-1.0.0-RC1(~Cf-2.0.0-RC1~Cf-2.2.0-RC1 ). With Leshan-1.0.0-M13(Cf-2.0.0) it seems this did not happen.

**Comment by boaks** [comment]
Now it makes sense again :-).

**Comment by boaks** [comment]
Do you think it's possible to test with a 2.2.2 (PR #1274 )? Or maybe you can cherry-pick that commit?

**Comment by sbernard31** [comment]
We face this only in production.

Currently in production we are no easy way to include patched version (cherry-pick) :/ and I'm not sure the deployment process could/should be modified in that way.

About a 2.2.2 with #1274, I don't know that means release a Leshan 1.0.0-RC2 with it ?
I read quickly the PR and I'm not sure to understand the impact.

About the NPE, have you any idea of the potential impact ?

**Comment by boaks** [comment]
> I read quickly the PR and I'm not sure to understand the impact.

It "hopefully" helps to limit the scope of the root cause.
It writes a log message as error, when the serial execution is violated.
If the NPE occurs without these messages, then I would guess, that there are some "duplicate messages" in the handshakers in buffer and would focus to search there.

**Comment by boaks** [comment]
I will anyway try to find some time over eastern, to check, if I see a relevant change since 2.0.0 and 2.2.0-RC1 . But I'm not sure, if that will result in findings. Sometimes "race-conditions" are "always" there, but just little change in the timing either trigger them or not. So such a search for THE change may be in vain.

**Comment by sbernard31** [comment]
I'm reading  `Handshaker.processMessage(Record)` and I see
```java
if (lastFlight) {
	// we already sent the last flight (including our FINISHED message),
	// but the other peer does not seem to have received it because we received
	// its finished message again, so we simply retransmit our last flight
```
but I see nothing in the code which checks we receive a **finished** message ? Or did I miss something ?

**Comment by boaks** [comment]
> but I see nothing in the code which checks we receive a finished message ?

If the last flight is sent, the last incoming  message was the finish.
The handshake state machine will filter other messages.

My current test points to a "close notify" instead of application data.
That processing of that was changed. My current assumption is, that check in
[handshake failed](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java#L1269) blocks the [cleanup for "close notify"](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java#L1525).

**Comment by boaks** [comment]
For now, the only way I could reproduce that NPE seems to be a `ResumingClientHandshaker` and a "close notify".

My current tests also uncovered a nasty race condition:
Peer send it's last app data and a close.
If that is processed in the reverse order, the connection is closed and the app-data is dropped.
To prevent from that, either the appdata and close notify must be put into one datagram, or scandium would require a redesign :-).

**Comment by boaks** [comment]
A "cheap cheat" maybe, not to CLOSE, just mark the session with the record_sequence_Number. Drop messages after that, process messages before that.

My feeling is:
Accept that race-drop for 2.2.x
Try the "cheap cheat" on 2.3.0

**Comment by boaks** [comment]
For the NPE itself, a experimental fix seems to work. I will prepare a PR for it.

**Comment by boaks** [comment]
My analysis results:
The handshaker reaches the "lastFlight" state only when the session was established and the handshaker waits for the next successful received record. 
If that record is a "close notify", the "handshakeFailed" may leave the handshaker in an unexpected state, with not fight but the lastFlight still at true. When that happens exactly, is so far out of my knowledge. Beside of the exception in the log, I didn't find some state to check for unit tests. 

To fix this cause, I provided a PR, which in case of "close_notify" completes the handshake, and for other alerts fails the handshake. To ensure, that the states lastFlight and flights to not diverge, the "handshakeFailed" does not longer consider, if the session is established or not.

**Comment by boaks** [comment]
The early "cheap cheat" is now in PR #1279 .

**Comment by sbernard31** [comment]
With Cf-2.2.2 we don't see this NPE anymore.

**Comment by sbernard31** [comment]
Still no NPE and I activate DEBUG log for few minutes and didn't see any related one.
So we can close this.

**Comment by sverhagen** [comment]
FYI, the extra logging to analyze this issue is still left-behind, and it's currently the only logger named differently than all the others:

    private static final Logger LOGGER_OWNER = LoggerFactory.getLogger(Connection.class + "owner");

Gives a logger named:

    class org.eclipse.californium.scandium.dtls.Connectionowner

(Particularly note the `class` in front, because `Connection.class + <string>` uses `Class.toString` which includes that prefix.)

Probably rather want this to be something like:

    private static final Logger LOGGER_OWNER = LoggerFactory.getLogger(Connection.class.getName() + ".owner");

Would give a logger named:

    org.eclipse.californium.scandium.dtls.Connection.owner

(Particularly note the lack of `class` in front.)

**Comment by boaks** [comment]
Thanks! Fixed on master.
@sverhagen do you need the fix also in a 2.3.2 or 2.4.2?
Currently I wouldn't plan a bugfix release for that, but, if you benefit from it, it's possible.

**Comment by sverhagen** [comment]
It was just something I noticed, but it breaks exactly nothing, so `master` is totally fine :+1:
